codespan-reporting/src/term/utils.rs

```diff
@@ -0,0 +1,102 @@
+use std::ops::Range;
+/*
+pub(crate) fn closest_char_boundary(s: &str, byte_index: usize) -> usize {
+    let length = s.len();
+    if let Some(index) = (byte_index..=length)
+        .skip_while(|&i| !s.is_char_boundary(i))
+        .next()
+    {
+        index
+    } else {
+        length
+    }
+}
+
+pub(crate) fn closest_char_boundary_rev(s: &str, byte_index: usize) -> usize {
+    if let Some(index) = (0..=byte_index)
+        .rev()
+        .skip_while(|&i| !s.is_char_boundary(i))
+        .next()
+    {
+        index
+    } else {
+        0
+    }
+}
+*/
+
+pub(crate) fn closest_char_boundary(s: &str, byte_index: usize) -> usize {
+    let length = s.len();
+    for index in byte_index..=length {
+        if s.is_char_boundary(index) {
+            return index;
+        }
+    }
+    length
+}
+
+pub(crate) fn closest_char_boundary_rev(s: &str, byte_index: usize) -> usize {
+    for index in (0..=byte_index).rev() {
+        if s.is_char_boundary(index) {
+            return index;
+        }
+    }
+    0
+}
+
+pub(crate) fn slice_at_char_boundaries<'a>(s: &'a str, range: Range<usize>) -> &'a str {
```

Comment:
I think `pub` would suffice here seeing as the module isn't publicly exported - I try to avoid `pub(crate)` - not the biggest fan of it. 😅